src/ir/analysis/derive_copy.rs

```diff
@@ -0,0 +1,416 @@
+            EdgeKind::InnerVar |
+            EdgeKind::Method => false,
+            EdgeKind::Generic => false,
+        }
+    }
+
+    fn insert(&mut self, id: ItemId) -> ConstrainResult {
+        trace!("inserting {:?} into the cannot_derive_copy set", id);
+
+        let was_not_already_in_set = self.cannot_derive_copy.insert(id);
+        assert!(
+            was_not_already_in_set,
+            "We shouldn't try and insert {:?} twice because if it was \
+             already in the set, `constrain` should have exited early.",
+            id
+        );
+
+        ConstrainResult::Changed
+    }
+
+    fn insert_array(&mut self, id: ItemId) -> ConstrainResult {
+        trace!("inserting {:?} into the cannot_derive_copy_in_array set", id);
+
+        let was_not_already_in_set = self.cannot_derive_copy_in_array.insert(id);
+        assert!(
+            was_not_already_in_set,
+            "We shouldn't try and insert {:?} twice because if it was \
+             already in the set, `constrain` should have exited early.",
+            id
+        );
+
+        ConstrainResult::Changed
+    }
+}
+
+impl<'ctx, 'gen> MonotoneFramework for CannotDeriveCopy<'ctx, 'gen> {
+    type Node = ItemId;
+    type Extra = &'ctx BindgenContext<'gen>;
+    type Output = (HashSet<ItemId>, HashSet<ItemId>);
+
+    fn new(ctx: &'ctx BindgenContext<'gen>) -> CannotDeriveCopy<'ctx, 'gen> {
+        let cannot_derive_copy = HashSet::new();
+        let cannot_derive_copy_in_array = HashSet::new();
+        let dependencies = generate_dependencies(ctx, Self::consider_edge);
+
+        CannotDeriveCopy {
+            ctx,
+            cannot_derive_copy,
+            cannot_derive_copy_in_array,
+            dependencies,
+        }
+    }
+
+    fn initial_worklist(&self) -> Vec<ItemId> {
+        self.ctx.whitelisted_items().iter().cloned().collect()
+    }
+
+    fn constrain(&mut self, id: ItemId) -> ConstrainResult {
+        trace!("constrain: {:?}", id);
+
+        if self.cannot_derive_copy.contains(&id) || self.cannot_derive_copy_in_array.contains(&id) {
+            trace!("    already know it cannot derive Copy");
+            return ConstrainResult::Same;
+        }
+
+        let item = self.ctx.resolve_item(id);
+        let ty = match item.as_type() {
+            Some(ty) => ty,
+            None => {
+                trace!("    not a type; ignoring");
+                return ConstrainResult::Same;
+            }
+        };
+
+        if item.is_opaque(self.ctx, &()) {
+            let layout_can_derive = ty.layout(self.ctx).map_or(true, |l| {
+                l.opaque().can_trivially_derive_copy()
+            });
+            return if layout_can_derive {
+                trace!("    we can trivially derive Copy for the layout");
+                ConstrainResult::Same
+            } else {
+                trace!("    we cannot derive Copy for the layout");
+                self.insert(id);
+                self.insert_array(id)
```

Comment:
Seems like having an `insert_both` method could be nice; these two lines are repeated in multiple places.